File: lc/coreprep/implementation/practice_iterations/two_pointers_1/PairWithTargetSum.java
import java.util.*;
import java.util.concurrent.*;
import java.util.function.*;

// O(n) / O(1)
public class PairWithTargetSum {
  public static void main(String[] args) {
    try (Scanner in = new Scanner(System.in)) {
      int tt = in.nextInt();

      while (tt-- > 0) {
        int n = in.nextInt();
        int s = in.nextInt();

        int[] a = new int[n];
        for (int i = 0; i < n; i++) {
          a[i] = in.nextInt();
        }

        int lpos = -1, rpos = -1;
        int left = 0, right = n - 1;

        while (left < right) {
          int sum = a[left] + a[right];

          if (sum == s) {
            lpos = left;
            rpos = right;
            break;
          } else if (sum < s) {
            left++;
          } else {
            right--;
          }
        }

        System.out.printf("%d %d\n", lpos, rpos);
      }
    }
  }
}

File: lc/coreprep/implementation/practice_iterations/iteration_archives/two_pointers_2/pair_with_target_sum.cpp
#include <algorithm>
#include <iostream>
#include <vector>

using namespace std;

// O(n) / o(1)
int main() {
  ios::sync_with_stdio(0);
  cin.tie(0);

  int tt, n, s;
  cin >> tt;

  while (tt--) {
    cin >> n >> s;

    vector<int> a(n);
    for (int i = 0; i < n; i++) {
      cin >> a[i];
    }

    int left = 0, right = n - 1;
    int lpos = -1, rpos = -1;

    while (left < right) {
      int sum = a[left] + a[right];

      if (sum == s) {
        lpos = left;
        rpos = right;
        break;
      } else if (sum < s) {
        left++;
      } else {
        right--;
      }
    }

    cout << lpos << " " << rpos << "\n";
  }

  return 0;
}